engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/ProcessInstanceStateTransitionGuard.java

```diff
@@ -271,4 +255,130 @@ private Either<String, ElementInstance> hasNonInterruptedFlowScope(
                   BufferUtil.bufferAsString(element.getId())));
     }
   }
+
+  private boolean doesPathExist(
+      final List<ElementInstance> instanceList, final ExecutableFlowNode toElement) {
+    final boolean doesExistPath = false;
+    int exclusiveCount = 0;
+    int checkedExclusiveCount = 0;
+
+    int boundaryCount = 0;
+    int checkedBoundaryCount = 0;
+
+    for (final ElementInstance instance : instanceList) {
```

Comment:
❌ I don't really understand your `doesPathExist` implementation. It seems to count something about exclusive gateways and boundary events, but that should not be necessary. More importantly, I think it cannot handle cycles (or loops) in the path where elements are visited multiple times. A typical example often used in BPMN processes is a retry loop.

So, how can we deal with cyclic paths?

A BPMN process can be considered a [directed graph](https://en.wikipedia.org/wiki/Directed_graph), where each element is a node in the graph, and each outgoing sequence flow is a directed edge (with some additional remarks I'll come back to).

Efficient algorithms exist to find paths in directed graphs. We can use such a path-finding algorithm for a graph data structure to find whether a path exists in an efficient way. The most common choice for that would be [Depth-First Search](https://en.wikipedia.org/wiki/Depth-first_search) (DFS), which keeps track of the visited elements and avoids visiting cyclic paths more than once. As soon as the DFS algorithm finds a path it could immediately return that a path is found, and so we have a rather efficient algorithm to determine whether a path exists between an active element instance and the inclusive gateway.

As I mentioned, a BPMN process can be represented as a graph. To be a bit more exact, I think we should consider the following:
- all elements are nodes
- sequence flows are directed edges between nodes, from the source element and to the target element
- container elements (e.g. [embedded subprocesses](https://docs.camunda.io/docs/components/modeler/bpmn/embedded-subprocesses/)) are simultaneously a node in the directed graph, but also contain a directed graph themselves
- [boundary events](https://docs.camunda.io/docs/components/modeler/bpmn/events/#boundary-events) can be considered a combination of a directed edge, from the element it is attached to, and a node, as the target of that edge, at the same time.
- [event subprocesses](https://docs.camunda.io/docs/components/modeler/bpmn/event-subprocesses/) can similarly be considered a combination of a directed edge, from the element it exists in (as it can be triggered while that element is active), and a node, as the target of that edge, at the same time; **and** it also contains a directed graph itself.